### ISSUE TITLE
WebSocketConnectionClosedException every 2-3 minutes since upgrade to Weechat 2.4

### ISSUE THREAD

**Issue by apetresc** [wee-slack/wee-slack#679]
I've been using wee-slack without issue for about a year, but since a few days ago when I upgraded to weechat 2.4, every 2-3 minutes I get reconnected to the server. When turning on debug mode, I see the following stack trace repeated ad nauseam:

```
08:47:30 | DEBUG: receive_ws_callback failed with exception:
08:47:30 | Traceback (most recent call last):
08:47:30 |   File "/home/apetresc/.weechat/python/autoload/wee_slack.py", line 391, in receive_ws_callback
08:47:30 |     data = team.ws.recv()
08:47:30 |   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 313, in recv
08:47:30 |     opcode, data = self.recv_data()
08:47:30 |   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 330, in recv_data
08:47:30 |     opcode, frame = self.recv_data_frame(control_frame)  
08:47:30 |   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 343, in recv_data_frame
08:47:30 |     frame = self.recv_frame()
08:47:30 |   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 377, in recv_frame
08:47:30 |     return self.frame_buffer.recv_frame()
08:47:30 |   File "/usr/lib/python2.7/site-packages/websocket/_abnf.py", line 361, in recv_frame
08:47:30 |     self.recv_header()
08:47:30 |   File "/usr/lib/python2.7/site-packages/websocket/_abnf.py", line 309, in recv_header
08:47:30 |     header = self.recv_strict(2)
08:47:30 |   File "/usr/lib/python2.7/site-packages/websocket/_abnf.py", line 396, in recv_strict
08:47:30 |     bytes_ = self.recv(min(16384, shortage))
08:47:30 |   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 452, in _recv
08:47:30 |     return recv(self.sock, bufsize)
08:47:30 |   File "/usr/lib/python2.7/site-packages/websocket/_socket.py", line 112, in recv
08:47:30 |     "Connection is already closed.")
08:47:30 | WebSocketConnectionClosedException: Connection is already closed.
08:50:53 | DEBUG: receive_ws_callback failed with exception:
08:50:53 | Traceback (most recent call last):
08:50:53 |   File "/home/apetresc/.weechat/python/autoload/wee_slack.py", line 391, in receive_ws_callback
08:50:53 |     data = team.ws.recv()
08:50:53 |   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 313, in recv
08:50:53 |     opcode, data = self.recv_data()
08:50:53 |   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 330, in recv_data
08:50:53 |     opcode, frame = self.recv_data_frame(control_frame)
08:50:53 |   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 343, in recv_data_frame
08:50:53 |     frame = self.recv_frame()
08:50:53 |   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 377, in recv_frame
08:50:53 |     return self.frame_buffer.recv_frame()
08:50:53 |   File "/usr/lib/python2.7/site-packages/websocket/_abnf.py", line 361, in recv_frame
08:50:53 |     self.recv_header()
08:50:53 |   File "/usr/lib/python2.7/site-packages/websocket/_abnf.py", line 309, in recv_header
08:50:53 |     header = self.recv_strict(2)
08:50:53 |   File "/usr/lib/python2.7/site-packages/websocket/_abnf.py", line 396, in recv_strict
08:50:53 |     bytes_ = self.recv(min(16384, shortage))
08:50:53 |   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 452, in _recv
08:50:53 |     return recv(self.sock, bufsize)
08:50:53 |   File "/usr/lib/python2.7/site-packages/websocket/_socket.py", line 112, in recv
08:50:53 |     "Connection is already closed.")
08:50:53 | WebSocketConnectionClosedException: Connection is already closed.
```

(Only pasted two instances here but if I leave it running overnight I wake up with literally hundreds of them)

I'm on Arch Linux, weechat 2.4, websocket_client 0.55.0. I was originally on the latest release of wee-slack but now I'm on `master` just to see if any of the latest commits had fixed anything. Let me know if there's any other information I can provide :)

(I should note wee-slack is otherwise usable if I just ignore the constant reconnects. It's just annoying)

**Comment by tardypad** [comment]
I have the same issue (same distro and package versions with wee-slack 2.2.0)
Well I didn't turn on the debug mode to check the details but the issue looks the same: I get reconnected every few minutes but everything looks fine besides it.
Just saying this so that it probably rules out a specific configuration issue on your environment.

**Comment by zingyb** [comment]
For me it's actually impeding the ability to use wee-slack. I can barely do anything before getting booted off

```
09:50 snip.slack.com     │ Lost connection to slack team snip.slack.com (on receive), reconnecting.
09:50 snip.slack.com     │ Lost connection to slack team snip.slack.com (on send), reconnecting.
```

This happens frequently enough that I get hit with ratelimits so my logs get spammed with `ERROR: Failed connecting to Slack with token starting with xoxp-snip: ratelimited`

I'm on up-to-date Arch Linux with the latest master of wee-slack and python-websocket-client 0.55.0-1

**Comment by celsworth** [comment]
I believe this may be related to the Slack API now requiring pings on the web socket.

With full credit to https://github.com/slackapi/python-slackclient/issues/118#issuecomment-348731791 for the idea, I went ahead and bodged in a ping function to my `wee-slack.py`:

```python
def slack_ping(a, b):
    for t in EVENTROUTER.teams.values():
        request = {"type": "ping"}
        t.send_to_websocket(request, expect_reply=False)
    return w.WEECHAT_RC_OK
```

(it seemed to want two args and it was quicker to ignore them than try to work out where they came from).

And then hooked it in every 5 seconds in `setup_hooks`:

```python
    w.hook_timer(5000, 0, 0, "slack_ping", "")
```

I've not had a disconnect for over 10 minutes with this running (was previously getting them every 3 minutes).

I think 5 seconds is probably appropriate, the [Slack API docs](https://api.slack.com/rtm) say:

> When there is no other activity clients should send a ping every few seconds

**Comment by zingyb** [comment]
Hm, this doesn't fix it for me. Perhaps I have a separate issue.

**Comment by immae** [comment]
Hey there,
I reproduced the error with websocket_client 0.55.0 . If you downgrade to 0.54.0, it works fine with weechat 2.4 and wee-slack any version (master of official release).

Note that it’s strange that websocket_client is out in pypi.org while it doesn’t exist on the maintainer’s github project (where it stops at 0.54.0)

Hope it helps,

**Comment by zingyb** [comment]
Still no good for me, either. Considering this happens almost immediately on script load and never stops, I'm just gonna say my issue is separate. Thanks for the help, though :(

**Comment by immae** [comment]
Hmm I tried every combination between weechat 2.3 - 2.4, wee-slack versions and websocket_client 0.54 - 0.55, and the websocket_client version is the only one in which I get the error (after a few minutes), so probably your issue is different yes.

Could you try another key maybe? (for instance, taking the one used in the web application)

**Comment by trygveaa** [comment]
@celsworth and @immae: Thanks for looking into this. So, it seems this could be caused by two different things. Either Slack requiring pings, or a change in websocket-client. I added pinging now, but since Slack apparently only require pings for certain teams, I think most people experience the problem because of an issue in websocket-client.

I figured out that the issue in websocket-client is caused by https://github.com/websocket-client/websocket-client/pull/474, which causes websocket-client to drop the connection when SSLWantReadError is thrown. I added a comment at https://github.com/websocket-client/websocket-client/pull/474#issuecomment-472348867. This patch fixes the issue:

```
diff --git a/websocket/_socket.py b/websocket/_socket.py
index d811c21..7087f1f 100644
--- a/websocket/_socket.py
+++ b/websocket/_socket.py
@@ -82,8 +82,6 @@ def recv(sock, bufsize):
     def _recv():
         try:
             return sock.recv(bufsize)
-        except SSLWantReadError:
-            pass
         except socket.error as exc:
             error_code = extract_error_code(exc)
             if error_code is None:
```

This essentially reverts a part of that PR though, so I'll have to look a bit more into what a proper fix would be (I'm not sure what issue that PR is fixing).

**Comment by ainola** [comment]
I'm still having this issue even with that latest master commit as well as manually removing those two lines from /usr/lib/python3.7/site-packages/websocket/_socket.py.

I'm hesitant to open a new bug report because it does seem so similar to the OP's issue:

```
21:10 │ Traceback (most recent call last):
21:10 │   File "/home/ainola/dotfiles/weechat/.config/weechat/python/autoload/wee_slack.py", line 391, in receive_ws_callback
21:10 │     data = team.ws.recv()
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 313, in recv
21:10 │     opcode, data = self.recv_data()
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 330, in recv_data
21:10 │     opcode, frame = self.recv_data_frame(control_frame)
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 343, in recv_data_frame
21:10 │     frame = self.recv_frame()
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 377, in recv_frame
21:10 │     return self.frame_buffer.recv_frame()
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_abnf.py", line 375, in recv_frame
21:10 │     payload = self.recv_strict(length)
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_abnf.py", line 396, in recv_strict
21:10 │     bytes_ = self.recv(min(16384, shortage))
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 452, in _recv
21:10 │     return recv(self.sock, bufsize)
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_socket.py", line 112, in recv
21:10 │     "Connection is already closed.")
21:10 │ WebSocketConnectionClosedException: Connection is already closed.
21:10 │ DEBUG: send_to_websocket failed with data: `{"type": "presence_sub", "ids": ["U0JHJM2G0", "UDW0JCUH2", "UEDCU7M0V", "U4NSX05GF", "UCWCSPQC8",
      │ "U9PUPJT62", "U0K0CD3EH", "U3T5524HH", "U2GGY953J", "UG9M8NERX", "U6XR50CJF", "U5LJKB1CZ", "U1EDNEMNU", "U3C5HDTQQ", "U1X286JEN", "U7D3A3473",
      │ "U7X841C2Y", "U9J2LB66R", "UEW8VKB40", "U9P34U4LC", "U93LMLR41", "U0JHYGTL7", "UCYKZCT4P", "U8M2GUVMH", "U8V64UUVC", "U9F41SB9A", "UGLSCRZK6",
      │ "U0JR9D405", "U3T7S6AUA", "UA8QA2QUU", "U0LN3JZB5", "U9CPLL5PT", "UC63LMKJT", "U4WEK0NUR", "UFLAH6V1D", "U2ACX8NCU", "UAD9APWEN", "U0JBVUH5Y",
      │ "U6YF1T553", "U2DCZ20LV", "UAEFRFREU", "UAUSBA5MJ", "UGUMF24SV", "U74SVJL7M", "U0JC50XNG", "USLACKBOT", "UBSD34SUV", "U8P4R5U8H", "U2PKQ1Q3X",
      │ "U0JP8G7S6", "U9LT4Q7HT", "U7P84CF3K", "U0JFXFYKZ", "U54G4A98V", "U3BPJG53M", "U9LNG71M0", "U5MM3BH9T", "UC9PZ1491", "U3VSMGVK3", "U0JHEQGTF",
      │ "U8WJRLFT4", "U4NMXSYUA", "U5PFQA59A", "U878QNYTD", "U93T60F1C", "UBC7MTUDS", "U4MAZP8Q1", "UBBGURX0W", "U6X0F5YP7", "UDW7K8D41", "U8M5U65NG",
      │ "U0JBU8UL8", "U0KQ0LK6U", "U5NDQAVU6", "U0RT3D2H2", "U4P4JF8BU", "U6Z7ZDRS8", "UCNCK4ZKM", "U0TCS6TUG", "UCX3XSJ0P", "U3TSYQP0V", "U52BHQVT3",
      │ "U8344BRSQ", "UAZM9NR9P", "U1PCXKF4M", "U4XTAGZN2", "UGFDV29NK", "U3H1G9NR3", "U5V0WAALU", "U0JHL35QQ", "U4N0D3G5Q", "U0JJ15607", "U5GFVBZHT",
      │ "U52S5A0NP", "U1KAJ94LT", "U8VK8MZNE", "U62QXDLBF", "UCNFZTSEN", "UCD0KKMM4", "U8JQ8FAV6", "UAUBAB1HC", "U5N2FUZNF", "U6A6LQ4F6", "UA1R4GZEF",
      │ "U9KJZ7A6M", "U89EBPGNT", "UFW0VQLLV", "U0JHYK329", "UAUA5NRR9", "U8KD89EQL", "U8JNYVBUY", "U2DNYRX89", "U3G01N7BJ", "UGUG7HGGH", "U7CJBM6LB",
      │ "U8L4A7S8J", "U0VLY364B", "U8X9N3TQB", "U3T91N64U", "UE8QLRKS9", "U0THWPYPP", "U43MSJ6U8", "U7WD49C1Y", "U8NC9HBTL", "U9B1EFKB6", "U845DL7B9",
      │ "U8TKR1CUQ", "U5DJNNF33", "UF2GMK57G", "U72LSGXR7", "UCP5Z8K8W", "U1D2ZURJR", "UB72T9ADR", "U4QHAUW3W", "U8X9LTA02", "UDF75G5GU", "U9DGK8LF8",
      │ "UGXCBNW4B", "U8GCMSAA0", "U0JPDBG3X", "U5FN12F5E", "U8RGV1V8S", "U37PBPB0S", "U0JBXM5AS", "U7TUGPV51", "U41H0MUBU", "U4G3M32MA", "U6YGPKKU3",
      │ "U6R8WJ5B3"], "id": 6}` and exception:
21:10 │ Traceback (most recent call last):
21:10 │   File "/home/ainola/dotfiles/weechat/.config/weechat/python/autoload/wee_slack.py", line 1237, in send_to_websocket
21:10 │     self.ws.send(encode_to_utf8(message))
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 253, in send
21:10 │     return self.send_frame(frame)
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 278, in send_frame
21:10 │     l = self._send(data)
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_core.py", line 448, in _send
21:10 │     return send(self.sock, data)
21:10 │   File "/usr/lib/python2.7/site-packages/websocket/_socket.py", line 132, in send
21:10 │     raise WebSocketConnectionClosedException("socket is already closed.")
21:10 │ WebSocketConnectionClosedException: socket is already closed.
```

Is there any more information I can provide?

**Comment by ainola** [comment]
Ah, I'm stupid; I edited the python 3 version. It appears that editing the python 2 version fixes it.

So in my case, the latest master commit for wee_slack.py fixed the insane amount of reconnects until I get rate-limited, and editing /usr/lib/python2.7/site-packages/websocket/_socket.py to remove that `SSLWantReadError` fixed the slower disconnects as described by the OP.

**Comment by apetresc** [comment]
For me, 55dfe74 actually increased the problem, I started getting disconnected every 5 seconds. But then I applied the patch to my system's `_socket.py` and that resolved the problem :) So I guess there were two distinct bugs here, one was fixed in master and the other has to be fixed upstream (or just revert to 0.54.0)

@trygveaa Should I close this issue, since technically my problem is (personally) resolved, or leave it open to track websocket_client's progress on a fix? (Judging by IRC recently it might be a good idea to leave it up for now)

**Comment by dbarrosop** [comment]
Same here, master fixed my disconnects, which then triggered "Connection already closed" and "ratelimit" errors and then downgrading to websocket-client 0.54.0 fixed those errors.

**Comment by trygveaa** [comment]
This has been fixed in websocket-client version 0.56.0.
